mypy/test/testsamples.py

```diff
@@ -1,31 +1,57 @@
 """Self check mypy package"""
 import sys
 import os.path
-from typing import List, Set
+from typing import List, Optional, Set
 
 from mypy.test.helpers import Suite, run_mypy
 
 
-class SamplesSuite(Suite):
-    def test_stubs(self) -> None:
-        # We only test each module in the one version mypy prefers to find.
-        # TODO: test stubs for other versions, especially Python 2 stubs.
-        seen = set()  # type: Set[str]
-        modules = []
-        # TODO: This should also test Python 2, and pass pyversion accordingly.
-        for version in ["2and3", "3", "3.5"]:
-            # FIX: remove 'builtins', this directory does not exist
-            for stub_type in ['builtins', 'stdlib', 'third_party']:
-                stubdir = os.path.join('typeshed', stub_type, version)
-                for f in find_files(stubdir, suffix='.pyi'):
-                    module = file_to_module(f[len(stubdir) + 1:])
-                    if module not in seen:
-                        seen.add(module)
-                        modules.extend(['-m', module])
+class TypeshedSuite(Suite):
+    def check_stubs(self, version: str, name: Optional[str] = None) -> None:
+        if name is None:
+            name = version
+        seen = {'__builtin__', }
+        modules = ['__builtin__']
+        for stub_type in ['stdlib', 'third_party']:
+            stubdir = os.path.join('typeshed', stub_type, name)
+            for f in find_files(stubdir, suffix='.pyi'):
+                module = file_to_module(f[len(stubdir) + 1:])
+                if module not in seen:
+                    seen.add(module)
+                    modules.extend(['-m', module])
+        if version == "3":
+            version = "3.4"
         if modules:
-            # these require at least 3.5 otherwise it will fail trying to import zipapp
-            run_mypy(['--python-version=3.5'] + modules)
+            modules.remove('__builtin__')
+            run_mypy(['--python-version={}'.format(version)] + modules)
+
+    def test_2(self) -> None:
+        self.check_stubs("2.7", name="2")
+
+    def test_2and3_2(self) -> None:
+        self.check_stubs("2.7", name="2and3")
+
+    def test_2and3_3(self) -> None:
+        sys_ver_str = '.'.join(map(str, sys.version_info[:2]))
+        self.check_stubs(sys_ver_str, name="2and3")
```

Comment:
Rather then checking the "2and3" directory in only these two cases, would it maybe make sense to pass in a list of directories we want to check into `check_stubs`?

E.g. we could do `self.check_stubs(["2.7", "2and3"])` inside `test_2`, do `self.check_stubs(["2and3", "3", "3.5"])` inside `test_35`...

I think this would probably make our tests more thorough, though the tradeoff is that each individual test would take longer.

(This is just an idea though -- maybe the tradeoff isn't worth it?)

Reply:
Part of the reason I did this was we already run tests in a matrix of options on Travis, so we'd end up testing all the versions of Python we support if we use the system version info. I collapsed the 2/2and3 and 3/2and3 tests together, but I think otherwise, there isn't a point in adding 3/2and3 to each test.